python/cudf/cudf/core/dataframe.py

```diff
@@ -6516,30 +6540,34 @@ def _apply_support_method(self, method, axis=0, *args, **kwargs):
 
             result = getattr(cupy, method)(arr, axis=1, **kwargs)
 
+            if is_datetime_dtype(common_dtype):
+                # Convert the values back, based on original time unit
+                self = self.astype(original_dtypes, copy=False)
```

Comment:
I'm not seeing where `self` is being modified. Rather than taking this approach I think it would be cleaner and safer to make a shallow copy of `self` where needed and avoid having to rely on mutating `self` back to the original state.

If this function errors and throws an exception in between modifying `self` and this point, we leave the user in an unexpected state.

Reply:
Really good point there. I'm thinking however the opposite (from consistency point of view). For example we have a dataframe:
```
DataFrame with 2 columns: .dtypes = [M8[ms], M8[ns]]
```
In the perfect world:
```
Step 1: convert to `common_type`: dtypes=[M8[ns], M8[ns]], values base unit = [M8[ns], M8[ns]]
Step 2: row wise op in cupy...
Step 3: convert back:  dtypes=[M8[ms], M8[ns]], values base unit = [M8[ms], M8[ns]]
```
If something goes wrong, in the shallow copy scenario:
```
Step 1: Make shallow copy: self_copy = self.copy(deep=False)
Step 2: Convert to `common_type`: self_copy.dtypes=[M8[ns], M8[ns]], values base unit = [M8[ns], M8[ns]]
Step 3: error from cuPy!
```
Users would be left with inconsistent state where `self.dtypes=[M8[ms], M8[ns]]` and `values base unit=[M8[ns], M8[ns]]`
But in inplace manner:
```
Step 1: Convert to `common_type`: self.dtypes=[M8[ns], M8[ns]], values base unit = [M8[ns], M8[ns]]
Step 2: error from cuPy!
```
Users would still have consistent state: `self.dtypes=[M8[ns], M8[ns]]` and `values base unit=[M8[ns], M8[ns]]`. It will be unexpected to the user that the `dtypes` are different from how they passed in, however.

Reply:
We could handle that unexpectedness also. Like catching the errors from the cupy codes and revert the value conversion, then explicitly throw the errors.

Reply:
In my mind if we error in trying to generate the `.values` it throws as opposed to falling back somewhere to return incorrect types. My fear is something along the lines of:
```
df = cudf.DataFrame(...)
df_copy = df.copy(deep=False)
df.some_rowwise_op() # Throws a RuntimeError Exception

df.equals(df_copy) # False
```

Because the `some_rowwise_op` modifies `df` before an exception occurs before the code that reverts to original state.

Reply:
Perhaps I shouldn't convert the values inplace from the beginning? Instead if we use `astype(common_type, copy=True)`, then we don't have to bother modifying `df` at all.

Reply:
> Perhaps I shouldn't convert the values inplace from the beginning? Instead if we use `astype(common_type, copy=True)`, then we don't have to bother modifying `df` at all.

We don't want to unnecessarily deep copy the data though, that would be expensive. If we make a shallow copy of the input dataframe then we should be safe to mutate the DataFrame inplace as long as we're not mutating the memory of the Columns themselves.

Reply:
If the memory is mutated, we should make a deep copy to the data? Is there anyway to identify when `astype` will mutate column memory?

Reply:
`astype` never mutates memory. It returns the same memory when possible if `copy=False`. It's only if we actually modify the memory that we run into issues, i.e. by doing somethjing like a scatter: `my_col[0:4] = [1, 2, 3, 4, 5]`

Reply:
Thanks for the explanation! Follow up Q: could you give an example when returning the same memory is not possible if `copy=False`?

Reply:
`cudf.Series([1, 2, 3], dtype='int64').astype('float64', copy=False)`

Any time the actual data needs to be changed or the bitwidth of the type changes we return in new memory. The `astype` call will never mutate the input object, but can return a new object that references the same memory which if mutated will mutate the input object.